src/openrct2/actions/ScenerySetRestrictedAction.cpp

```diff
@@ -0,0 +1,63 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2024 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "ScenerySetRestrictedAction.h"
+
+#include "../object/SceneryGroupObject.h"
+#include "../world/Scenery.h"
+
+ScenerySetRestrictedAction::ScenerySetRestrictedAction(ObjectType objectType, ObjectEntryIndex entryIndex, bool isRestricted)
+    : _objectType(objectType)
+    , _objectIndex(entryIndex)
+    , _isRestricted(isRestricted)
+{
+}
+
+void ScenerySetRestrictedAction::Serialise(DataSerialiser& stream)
+{
+    GameAction::Serialise(stream);
+
+    stream << DS_TAG(_objectType) << DS_TAG(_objectIndex) << DS_TAG(_isRestricted);
+}
+
+void ScenerySetRestrictedAction::AcceptParameters(GameActionParameterVisitor& visitor)
+{
+    visitor.Visit("objectType", _objectType);
```

Comment:
@Gymnasiast we do not have access to the integers in the plugin API, only the strings. It may need a string-to-int mapping then in here? The table for that should already exist somewhere.

Reply:
I don’t think that is possible with the visitors. Maybe @duncanspumpkin can tell?

Reply:
You are correct that it doesn't support it nicely. You would have to have an intermediate string and then you need to magic the conversion function that is buried deep in the scripting headers.
```
std::string objectTypeStr;
visitor.Visit("objectType", objectTypeStr);
_objectType =  StringToObjectType(objectTypeStr);
```
But that has the issue that it only works one way and you need both ways. In conclusion. We can't fix this for now. We will have to change the .d.ts file to make this not an ObjectType but a number with a little blurb saying sorry.

Reply:
Would a custom `Visit()` overload be able to do it?

Otherwise I'd rather we leave the plugin part for another PR than to add a ready-to-deprecate implementation. For plugins it's just too messy to change things later because it will be able to break plugins that have been made in the mean-time.

Reply:
The Visit function has no concept of direction so no a custom overload wouldn't work. First you need to make Visit know of direction of use then you need to modify the DukTo/From functions to set the direction. Then you need to do the fiddly intermediate string to object thing. Unless @ZehMatt has a different idea? (The problem is we want to use a string type for plugins and an enum type for c++ code)

Okay sounds like we will have to remove the AcceptParameters for now and the .d.ts changes.

Reply:
What about an overload like this? Then it could have implementations for both DukTo and DukFrom akin to how it's done at other places where the maps are used bi-directional.
```
visitor.Visit("objectType", _objectType, StringEnumMapHere);
```
Would that be possible with maybe some template magic? (Could still be for another PR)

Reply:
> What about an overload like this? Then it could have implementations for both DukTo and DukFrom akin to how it's done at other places where the maps are used bi-directional.
> 
> ```c
> visitor.Visit("objectType", _objectType, StringEnumMapHere);
> ```
> 
> Would that be possible with maybe some template magic? (Could still be for another PR)

Hmm i guess that would work. I still think best to do as a separate PR.

Reply:
Not much, to be honest. I have had some trouble adding scripting tentacles in general, it’s not quite my strongest suit. Feel free to work on it yourself if you fancy.

Reply:
Does your original implementation still exist somewhere? We use magic numbers everywhere for the actions, so while it is not ideal, it is also not more complicated to use than some other actions (looking at you, TileModifyArgs!).